Add New JS Syntax Issue #2147

jsnorek opened the pull request:
This is a PR for issue #2147. The goal for this issue is to add new JS syntax to the given [topic outline](https://github.com/Techtonica/curriculum/blob/main/javascript/newer-JS-features.md).

I've made sure to keep updates consistent with the current lesson structure.
This updated file now includes new JS syntax such as Temporal API Features, Pipe Operator, and Records and Tuples, and gives information on the new feature as well as a code example. 

![Screenshot 2024-12-23 at 5 33 16 PM](https://github.com/user-attachments/assets/ee3150b9-774d-4fcb-87d5-a46bf39e4e2e)

I also updated the "test your knowledge" section to include some of the newly added features.

![Screenshot 2024-12-23 at 5 34 14 PM](https://github.com/user-attachments/assets/1620bbb5-79d0-4348-9d93-ac4c02a067e6)

Lastly, I added a few more relevant links to the "supplemental materials" section.

![Screenshot 2024-12-23 at 5 36 10 PM](https://github.com/user-attachments/assets/11d30785-152b-4bf6-b46d-05dae6a857dc)

daaimah123 commented on javascript/newer-JS-features.md:
```diff
@@ -222,6 +222,184 @@ const { email, password } = user; //equivalent to saying const email = user.emai
+
+Temporal API provides better date and time handling capabilities in JavaScript. We'll look at the classes: `PlainDate`, `PlainTime`, and `ZonedDateTime`.
+
+### **PlainDate**
```
```suggestion
##### **PlainDate**
```

jsnorek replied:
I've added this along with some additional formatting updates in a commit

daaimah123 replied:
It’s best practice to cite which commit by providing your hash in the replying comment.

daaimah123 commented on javascript/newer-JS-features.md:
````diff
@@ -222,6 +222,184 @@ const { email, password } = user; //equivalent to saying const email = user.emai
+console.log(date.day); // 25
+```
+
+### **PlainTime**
````
```suggestion
##### **PlainTime**
```

jsnorek replied:
I've added this along with some additional formatting updates in a commit

daaimah123 replied:
It’s best practice to cite which commit by providing your hash in the replying comment.

daaimah123 commented on javascript/newer-JS-features.md:
````diff
@@ -222,6 +222,184 @@ const { email, password } = user; //equivalent to saying const email = user.emai
+console.log(time.second); // 0
+```
+
+### **ZoneDateTime**
````
```suggestion
##### **ZoneDateTime**
```

jsnorek replied:
I've added this along with some additional formatting updates in a commit

daaimah123 replied:
It’s best practice to cite which commit by providing your hash in the replying comment.

daaimah123 reviewed (changes requested):
This looks good, it looks like some of the formatting needs to be consistent across all sections, and then I had a question about a bolding choice. Pending your incoming changes in response to this request for changes, this looks good 👏🏾

jsnorek commented on javascript/newer-JS-features.md:
````diff
@@ -247,6 +425,45 @@ const produce = [];
+console.log(now.toString());
+```
+
+1.	**“2024-12-23T10:35:00” (or your current date and time in ISO format).**
````
It's bolded because it's the answer to the multiple choice question. I followed the bolding format of the original file as the first question is multiple choice and the answer is bolded as well!

<img width="603" alt="Screenshot 2024-12-26 at 7 03 16 PM" src="https://github.com/user-attachments/assets/1dc9e83a-20e7-4b6e-8d2b-4e9e8cb7b943" />

jsnorek commented:
@daaimah123 I've updated the formatting requests and I also added the text "Example" above the code example in each new JS feature to maintain consistency 

![Screenshot 2024-12-26 at 7 09 41 PM](https://github.com/user-attachments/assets/a9e9f398-7e8a-4a58-8262-5f307e7ff536)

_____________
![Screenshot 2024-12-26 at 7 10 17 PM](https://github.com/user-attachments/assets/8a07c8f0-03fa-44e1-9cb7-4bfe85a660d9)

mai-repo reviewed (approved):
Hi @jsnorek , 

Everything looks great to me. If I could make a small suggestion, for the supplemental materials, I think it might be helpful to add the media type for the various links. For example:

- ES6 Tutorial (Video, 56mins)

Other than that -- it should be good to go! 👍

jsnorek commented:
@mai-repo I've added the media types to that list!

jsnorek commented:
Thanks! @daaimah123 @mai-repo After making the updates, I'm merging the branch 👍
